### ISSUE TITLE
Lambda shorthand -- _ placeholder character, and/or $1 variables

### ISSUE THREAD

**Issue by bandleader** [dotnet/vblang#256]
Express lambdas as "incomplete expressions," using a "placeholder character" (here, `_` as in Scala).
- Greatly improves readability, reducing the noise of `Function(parameter)`
- Reduces the need for for LINQ query comprehensions
- Well-received in Scala/Swift
- A perfect match for VB -- as it was decided that C# arrow syntax did not fit well into BASIC, whereas this one fits in very well
```vb
'Old way:
Dim activeCustomers = customers.Where(Function(c) c.LastActive > lastMonth).OrderBy(Function(c) c.LastName)
Dim lastNameList = activeCustomers.Select(Function(c) $"{c.LastName}, {c.FirstName}")
'New way:
Dim activeCustomers = customers.Where(_.Active).OrderBy(_.LastName)
Dim lastNameList = activeCustomers.Select($"{_.LastName}, {_.FirstName}")

'Here's how it works: 
'An expression that contains a reference to '_' is wrapped into a function.
Dim myLambda As Func(Of Customer, DateTime) = _.LastActive 
```
- Argument and return types can be inferred just as we do for the regular Function(args) syntax.

###  Multi-argument functions

Swift actually goes further and lets you use `$1`, `$2`, and so on, for lambdas where a multi-argument function is expected -- for instance, comparer/aggregator functions:
```vb
'Old way:
Dim totalOrders = activeCustomers.Aggregate(0, Function(runningSum, c) runningSum + c.Orders.Count)
'New way:
Dim totalOrders = activeCustomers.Aggregate(0, $1 + $2.Orders.Count)
```

### Partial application

This also shines for highly expressive partial application:
```vb
```

     

gets simplified to this:

     Dim FirstChar = Strings.Left(_, 1)

(In C# it wasn't so bad:

     var FirstChar = m => Strings.Left(m,1)

...but they the arrow function doesn't fit into BASIC -- well this one does nicely.)

BTW it doesn't just have to do lambdas, it's great in many places like in pattern matching:

Select Case variable

   Case 5: ...

   Case 10: ...

   Case _ Mod 3 = 4: ...

End Select

**Comment by franzalex** [comment]
I've always found VB.NET's syntax for creating anonymous methods (and lambda expressions) a tad too verbose. In that vein, I am all in favour of a simplified form as proposed here.

I just think that if considered for implementation, it should be the `$0`, `$1`, etc. form instead of the underscore form since the former is clearer to read than the latter.

**Comment by ericmutta** [comment]
Like @franzalex I understand very much the need for something like this: a less verbose way to write lambdas and support the overall idea 👍 

However, I think as far possible we should just harmonize with C# and use the `()=>` form which has the added benefit of being able to name your variables rather than using a punctuation character. So for example, instead of:

```vbnet
Dim lastNameList = activeCustomers.Select(Function(c) $"{c.LastName}, {c.FirstName}")
```
It would be:

```vbnet
Dim lastNameList = activeCustomers.Select((c)=> $"{c.LastName}, {c.FirstName}")
```

In essence we've just removed the `Function` keyword to reduce verbosity, but we can still keep the variable name `c` which clearly stands for 'customer' in that expression (using a symbol requires a mental "context switch" to figure out what's going on and feels a little too "magical").

Here are the other examples you give:

```vbnet
Dim isActiveLately As Func(Of Customer, DateTime) = (c) => c.LastActive > Now.AddMonths(-1)

Dim GetFirstChar = (str) => Strings.Left(str, 1)

Dim totalOrders = activeCustomers.Aggregate(0, (r, c) => r + c.Orders.Count)

Dim sortFunc = If(sortByTitle, (i As Issue) => i.Title, (i As Issue) => i.PostDate)
```

...seems to work for all cases and has the added bonus of being consistent with C# (which not an explicit goal, consistency does mean someone else already knows how to implement this syntax and that work can be leveraged for parsing, intellisense, etc).

**Comment by reduckted** [comment]
I'm all for reducing the verbosity of declaring lambdas!

I don't like the placeholder character. One problem with it is that it only works for single-parameter lambdas. If we're going to make lambdas less verbose, then we need something that works for _all_ types of lambdas.

Using a fat-arrow (`=>`) isn't too bad. You can name parameters and it works for any lambdas of any arity, though I fear it will be rejected because it's not "VB-like". 

Just to throw another idea into the mix, what about just dropping the `Function` or `Sub` keyword?

```vb
Dim isActiveLately As Func(Of Customer, DateTime) = (c) c.LastActive > Now.AddMonths(-1)

Dim GetFirstChar = (str) Strings.Left(str, 1)

Dim totalOrders = activeCustomers.Aggregate(0, (r, c) r + c.Orders.Count)

Dim sortFunc = If(sortByTitle, (i As Issue) i.Title, (i As Issue) i.PostDate)
``` 

I'm not overly fond of it, but maybe it will spur on some other great suggestions. 😄

**Comment by gilfusion** [comment]
If we went with ```()=>```, what would this do?
```vb
Dim myLambda = (ByRef x) => x = 5
```
Does it check if x is 5 and return a Boolean, or does it assign 5 to x?
Is there a way to keep the syntax compact, readable, and able to discern between those two possibilities?

**Comment by reduckted** [comment]
Hmm, yes, without a keyword you can't tell if its a `Function` or a `Sub`. I guess you'd (maybe) be able to work it out if it was passed as a parameter, but not if you're just assigning it to a value like that.

Maybe `=>` could only be a replacement for `Function`?

**Comment by AdamSpeight2008** [comment]
You can not express lambda with byref parameter(s) unless you state the target type, which tend to be a delegate.

```vbnet
Delegate Function TryParseDelegate(Of T)( Here As Int32, Diagnostics As DiagnosticBag, <Out>Byref Output As T ) As Boolean

Dim x As TryParseDelegate = ( x, d, byref o) => ...
```

**Comment by bandleader** [comment]
While the idea of adopting C#'s `=>` may well have merit, I feel this issue is getting hijacked 😃. This issue is not about shortening *regular* lambda syntax (argument list + expression/statement block), but about **omitting the argument list entirely** (in addition to the word Function) -- allowing you to succinctly express your logic without ANY other ceremony at all.
```vb
'Short and sweet, with nothing at all that you don't need. 
allCustomers.Where(_.Active).OrderBy(_.LastName) 

'A lot of ceremony just for specifying which property to use.
allCustomers.Where(Function(c) c.Active).OrderBy(Function(c) c.LastName) 
```

> @ericmutta However, I think as far possible we should just harmonize with C# and use the ()=> form which has the added benefit of being able to name your variables rather than using a punctuation character.

When you *want* to name your variables, that is what "regular" lambda syntax is for; this issue is about omitting the argument list. (As above, feel free to open an issue about shortening it and/or allowing `=>`.)

> @reduckted One problem with it is that it only works for single-parameter lambdas. If we're going to make lambdas less verbose, then we need something that works for all types of lambdas.

1. As noted above, the vast majority of the lambdas **targeted by this issue** are going to be single-argument lambdas. 
2. However, see the `$1` syntax above -- it works beautifully for multi-argument lambdas as well.

Thanks for your input! If anyone wants me to open a new issue about `=>` syntax for regular lambdas, let me know.

**Comment by gilfusion** [comment]
Another idea, from the keywords over punctuation department:

A ```Using``` expression (not completely thrilled reusing a keyword, but at least it's somewhat readable) that creates a single-parameter lambda whose parameter is accessed like in a ```With``` block.

### Examples
```VB
Dim activeCustomers = customers.Where(_.LastActive > lastMonth).OrderBy(Using .LastName)
Dim lastNameList = activeCustomers.Select(Using $"{.LastName}, {.FirstName}")
```

There is a pretty big difference between accessing a variable and creating a lambda expression, and this puts a brightly-colored (depending on your code editor settings) keyword there to remind you of that. It's shorter than the full ```Function(x)``` syntax while still being readable.

It doesn't quite catch all the cases above, but it does catch a few of them. (I could see maybe combining this syntax with the ```_``` or ```$1``` or something similar to catch the rest.)

**Comment by bandleader** [comment]
Found this at dotnet/csharplang#91
(I read recently that things common to both languages could/should be discussed on csharplang anyway)

@gilfusion I think I agree with the idea of having an explicit keyword/character to denote the start of the lambda. As per the discussion on csharplang, it would make parsing easier, and you also have a good point about making it obvious to humans.

If we go this way (which I don't mind) we aren't introducing "incomplete expression" syntax like Scala, but rather just allowing you to skip the parameter list, instead defining them as `$1`, `$2`, etc. as used in the lambda expression. (Or `_`)

**Point:** If instead of `Using` we kept `Function`, we'd be **very** close to existing VB lambda syntax:
```vb
'New way:
Dim lastNameList = activeCustomers.Select(Function $"{_.LastName}, {_.FirstName}")
'Old way:
Dim lastNameList = activeCustomers.Select(Function(c) $"{c.LastName}, {c.FirstName}")
```

...which is great because it makes it very easy to self-learn. It's pretty obvious what it does. 

The only thing is, that if we do use a keyword, `Function` is **way** too long, and takes up a large percentage of the expression. In fact, it retains more characters than it savces. Even `Using` is too long. See [this comment on csharplang](https://github.com/dotnet/csharplang/issues/91#issuecomment-361494339) which proposes `=>`.

BTW, using a keyword is similar to Swift's approach, where you still use braces to mark the start and end of the lambda: `{ $1.LastName }` -- you just use `$1` instead of an explicit parameter list. So if go this way, our feature shouldn't be called "incomplete expressions" but rather "shorthand lambda arguments."

I still like it and think we should do it. My favorite syntax thus far:
```vb
Dim activeCustomers = customers.Where(=> _.LastActive > lastMonth).OrderBy(=> _.LastName)
Dim lastNameList = activeCustomers.Select(=> $"{_.LastName}, {_.FirstName}")
```

P.S. We could handle multiple arguments with `_1` and `_2`. I think cleaner than `$1`, especially in conjuction with our syntax for string interpolation.
P.P.S. There isn't really a problem with `_1` etc. being existing ambiguous with a valid identifier, because if we're using a keyword anyway, `_1` is indeed an actual identifier implicitly defined within the scope of the method body, and isn't any different than using `Function(_1) _1.LastName` -- if you want to shadow an already-declared variable, go ahead (or we can make it illegal for clarity).

**Comment by KathleenDollard** [comment]
This is interesting. I have concerns about the brevity being clear, and there is a case where this may be ambiguous

``` VB
M2(M(_.Bar))
```

Is this 

```VB
M2(M(Function(x) x.Bar))
```

or 

```VB
M2(Function(x) M(x.Bar))
```

But even if that was resolved, I think the win in characters in this is fairly small and there is some loss in clarity.

**Comment by bandleader** [comment]
@KathleenDollard We can check how Scala decides this ambiguity, but usually there is no problem, because usually either M or M2 is expecting a lambda as a parameter, but not both.

And inspecting the signatures was already necessary here, so that the compiler can infer the type of `x`.

**Comment by Berrysoft** [comment]
How about the keyword `With`?
``` vbnet
Dim activeCustomers = customers.Where(With .LastActive > lastMonth).OrderBy(With .LastName)
```
It isn't ambiguous when
``` vbnet
'Old
M2(M(Function(x) x.Bar))
M2(Function(x) M(x.Bar))
'New
M2(M(With .Bar))
M2(With M(.Bar))
```
The space between `With` and dot is optional.

**Comment by bandleader** [comment]
@Berrysoft `With` isn't really that much shorter than Function (saves 4 chars). My main gain here was skipping the argument list, letting it be specified as `_` or as `$1`.

That said: if we absolutely do need a keyword for parsing reasons, we could even keep the word `Function` -- for me, **just making the argument list optional would already be a gain.** It's just more DRY:
```vb
'Old way:
Dim activeCustomers = customers.Where(Function(c) c.LastActive > lastMonth).OrderBy(Function(c) c.LastName)
Dim lastNameList = activeCustomers.Select(Function(c) $"{c.LastName}, {c.FirstName}")

'New way:
Dim activeCustomers = customers.Where(Function _.LastActive > lastMonth).OrderBy(Function 
 _.LastName)
Dim lastNameList = activeCustomers.Select(Function $"{_.LastName}, {_.FirstName}")
```
...but I really prefer the keyword shortened (perhaps `=>` as discussed above) or omitted altogether as in Scala.

**Comment by reduckted** [comment]
> `With` isn't really that much shorter than `Function` (saves 4 chars). 

> we could even keep the word `Function`

That's a bit contradictory, don't you think 😁 

One thing that I don't think has been mentioned is that whatever happens with `Function(x)`, it would be great to do the same thing (or something similar) with `Sub(x)`.

**Comment by bandleader** [comment]
> @reduckted That's a bit contradictory, don't you think 😁

No, read the words in between:

> That said: if we absolutely do need a keyword for parsing reasons,

**Comment by reduckted** [comment]
> ...it would be great to do the same thing (or something similar) with `Sub(x)`.

Having said that, `Sub` is already pretty short, so there's no great savings to be had there, though for the sake of consistency I think we'd want to do something with it.

> ```vb
> Dim activeCustomers = customers.Where(Function _.LastActive > lastMonth)
> ```

If we're going to use the `Function` keyword, there's really not a lot to be saved, which I think defeats the whole point of this request. 

Also, it's worth noting that using an underscore will (I'm pretty sure) break backward compatibility. As horrible as it is, this is valid code today:

```vb
Dim activeCustomers = customers.Where(Function _
    (x) x.LastActive > lastMonth)
```

**Comment by zspitz** [comment]
> Also, it's worth noting that using an underscore will (I'm pretty sure) break backward compatibility. As horrible as it is, this is valid code today:
>
>     Dim activeCustomers = customers.Where(Function _
>        (x) x.LastActive > lastMonth)

I think we could differentiate based on whether the `_` is followed by a line break -- then it's a line continuation -- or not, then it's the implicit variable.

**Comment by bandleader** [comment]
@zspitz @paul1956 To the best of my compiler parser knowledge, neither line continuation, nor comments after line continuation would present a problem here. Line continuation only occurs if `_` is the last character on the line (or before a `'` comment). 

(AFAIU the difficulty you speak of would be a problem in using _ at the beginning of an identifier as well, which is clearly allowed. / I defer to the Roslyn folks if I'm wrong.)

That said, I'm perfectly fine with using `$1` like Swift; in fact it's more flexible, as it can be used for multiple arguments (`$2` etc.). I just thought the dollar sign doesn't fit so nicely into BASIC. The `_` is very elegant as a symbol choice -- it looks like a "fill in the blank" sign, as in "place argument here." We could do `_1`, `_2` etc., but that loses a lot of the elegance IMHO.

Thank you everyone for your continued feedback. I really hope this gets into VB sometime soon. I don't see why we have to always be 2 versions behind C#, nor why some at MS fear that new users will be scared off by new features; on the contrary, most of our features add elegance and productivity (and are opt-in as well). VB should be _leading_ the languages community with its innovation and elegance.

**Comment by jrmoreno1** [comment]
I think a keyword is fine, I just think that Function is way to long.  While Fn may not be an English word, I think it's a well enough know concept and abbreviation to be worth using.  Fn(x) x is half as long as Function(x) x which really matters when you have fluent or nested lambda's.

**Comment by bandleader** [comment]
There is a Babel plugin to do this for JavaScript. The source code of the plugin looks very simple. Might be useful.

https://github.com/xtuc/babel-plugin-transform-scala-lambda/tree/master/src

**Comment by bandleader** [comment]
Also here's the part from the Scala spec which describes this -- though it's mathematical language.

https://www.scala-lang.org/files/archive/spec/2.11/06-expressions.html#placeholder-syntax-for-anonymous-functions

**Comment by bandleader** [comment]
Just discussed this with a Scala dev with some understanding of the compiler. He told me basically that this can be done completely at the parser level. It works like this:

1. "Underscore replacement" means taking an expression containing underscore identifiers `_`, switching each one out for a fresh token, and converting to an anonymous function. For example, `_.Trim()` gets converted into `Function(a) a.Trim()`, and `_ + _` gets converted into `Function(a, b) a +b`.
2. Underscore replacement is done when parsing _top-level_ expressions. However, this includes arguments (such as each of the two arguments in `func(_ + 2, _.Trim())`), and parenthesized expressions.
3. However, underscore replacement is _not_ done if the expression contains only an underscore token and nothing else -- in this case the parser leaves the token in, and the next-level-up expression will do the replacement. (i.e. in `array.Select(Foo(_))` doesn't convert to `.Select(Foo(Function(a) a))` but rather `.Select(Function(a) Foo(a))`.)

**Comment by jrmoreno1** [comment]
@bandleader when @zspitz said

> since `Fn` is not a recognizable English word, it's harder to parse this statement.

That was in reference to the programmer mentally parsing the statement, not the compiler.  Using symbols or spacing/positions makes that worse.

@KathleenDollard: in this case lack of brevity is harming clarity.  The length of the keyword and the rules for implied line continuation results in it being harder to read a chain of lambdas in VB and thus they get avoided. Basically it’s not too bad for a single predicate, but if you need more than one, there’s a good chance you’ll end up using some other syntax.

**Comment by zspitz** [comment]
@jrmoreno1 

> harder to read a chain of lambdas in VB

Presumably you mean something like this:
```vb
Dim lst = Enumerable.Range(1, 100).Select(Function(x) x*x).Where(Function(x) x Mod 2 = 0).OrderByDescending(Function(x) x).ToList
```
which is admittedly a little hard to follow. But we can always do this:
```vb
Dim lst = Enumerable.Range(1, 100).
    Select(Function(x) x*x).
    Where(Function(x) x Mod 2 = 0).
    OrderByDescending(Function(x) x).
    ToList
```
if not breaking each operation into a separate variable.

Also, note that the corresponding statement in C# is not much more readable:
```csharp
var lst = Enumerable.Range(1, 100).Select(x => x * x).Where(x => x % 2 == 0).OrderByDescending(x => x).ToList();
```
It's true that C# is more flexible about where you can put a line break, but that doesn't necessarily improve readability:
```csharp
var lst = Enumerable.Range(1, 100).Select(x => x 
    * x)
    .Where(x => 
    x % 2 == 0
    ).OrderByDescending(x => x).
    ToList();
```

**Comment by pricerc** [comment]
At risk of appearing to be trolling.

Wouldn't the VB way be:

```VB
    Dim lst2 = (
        From x In Enumerable.Range(1, 100)
        Where x Mod 2 = 0
        Select square = x * x
        Order By square Descending
    ).ToList()
```

?

I realise that lambdas are not just about LINQ, but it is the most common place where I see them discussed, I believe at least in part because C# chose not to implement LINQ keywords for 'trivial lambdas' (my interpretation). 

I do wonder if we'd be having this discussion if C# had implemented LINQ as fully as VB.

**Comment by bandleader** [comment]
And I always thought that if we only had the `_` suggestion (or similar), we wouldn't even need special query syntax:
```vb
Dim lst2 = (
  Enumerable.Range(1, 100)
  .Where(_ Mod 2 = 0)
  .Select(_ * _)
  .OrderByDescending(_)
).ToList()```

(Which of course would mean that with syntax nearly as clean as query syntax, we could do much more than just what query syntax supports.)

**Comment by pricerc** [comment]
> IMHO, a major reason they don't have LINQ query syntax as fully as we do, is because their 
> lambda syntax is much nicer, so they have much less need for it. I'm all for righting that wrong.

Yes, and no. 

I did read about it not long ago (not sure where) - basically, as I understand it, C# added keywords for the major concepts that they considered might be awkward or tricky to use as a lambda. And it was a conscious decision to not include those items with a 'trivial' lambda use.

Since VB has 'human-friendliness' as a major goal, they chose to implement much more in the language, and make it pretty much complete.

Which you prefer is, of course completely subjective, but the discussion I stumbled over was started by a question about why these things were missing from C#, which IMO has just a 'teaser' version of a LINQ syntax.

I prefer reading the LINQ syntax. I think it is easier to follow if you're looking at code again after months away, which I think is much more important that trying to make source code concise.

But I use both extensively, especially since it's generally much easier to translate C# examples into lambdas (and vice-versa, I suppose).

**Comment by zspitz** [comment]
@bandleader Note that your code doesn't compile, which is precisely what I was getting at. Because C# has an end-of-statement character (`;`), it can afford to be more flexible about line breaks; while VB.NET has to be more rigid about where the line breaks go. But that doesn't mean lines cannot be broken in VB.NET at all, and this provides the natural solution to the "chains of lambdas" that @jrmoreno1 mentioned.

**Comment by jrmoreno1** [comment]
@zspitz: actually I was thinking more of nested queries.  And yes both can be wrapped.  But methods with multiple loops and conditions, make even that problematic.

    Private Function Example(input as List(of Something, extraData as List(Of ItemswithList)) as List(Of SomethingElse)
         If (guardConditionNeedsReversal) Then
              If(seecondConditionThatNeedsReversed) Then
                  For Each a in input
                      If (a.whatever) Then
                           Dim lst = extraData.
                                           SelectMany(Function(x) x.ItemsList.
                                                      Select(Function(li) new With{PropertyA, PropertyB}).
                                                      Where(Function(a) test(a.PropertyA, a.PropertyB) >= constA).
                                           OrderByDescending(Function(x) x).
                                           ToList
                      End if
                  Next
              Else
                   Throw New Exception()
              End if
         Else 
             Throw New Exception()
         End if

    End Function

And that isn’t particularly deep in my experience.

**Comment by zspitz** [comment]
@jrmoreno1 I don't understand how your example demonstrates [this](https://github.com/dotnet/vblang/issues/256#issuecomment-461409813)

> The length of the keyword and the rules for implied line continuation results in it being harder to read a chain of lambdas in VB

Nested `If..Then..Else`, `For Each..Next` and conditions within `For Each` are hard to read in any language:
```csharp
List<SomethingElse> Example(List<Something> input, List<ItemswithList> extraData)
    if (guardConditionNeedsReversal) {
        if (seecondConditionThatNeedsReversed) {
            foreach (var a in input) {
                if (a.whatever) {
                    var lst = extraData
                        .SelectMany(x => x.ItemsList.
                            .Select(li => new {li.PropertyA, li.PropertyB})
                            .Where(a => test(a.PropertyA, a.PropertyB) >= constA)
                        )
                        .OrderByDescending(x => x)
                        .ToList();
                }
            }
        } else {
            throw new Exception();
        }
    } else {
        throw new Exception();
    }
}
```

but that has nothing to do with the length of VB's lambda syntax or line continuation rules; in spite of them, your example can be written far more simply:
```vb
Private Function Example(input as List(of Something, extraData as List(Of ItemswithList)) as List(Of SomethingElse)
    If Not guardConditionNeedsReversal Then Throw New Exception
    If Not seecondConditionThatNeedsReversed Then Throw New Exception 
    For Each a in input.Where(Function(x) x.whatever)
        Dim lst = extraData.
            SelectMany(Function(x) x.ItemsList.
                Select(Function(li) New With{PropertyA, PropertyB}).
                Where(Function(a) test(a.PropertyA, a.PropertyB) >= constA
            )).
            OrderByDescending(Function(x) x).
            ToList
    Next
End Function
```

**Comment by jrmoreno1** [comment]
@zspitz: absolutely it could be rewritten to be better (that was the entire point of the name I gave the conditions), but the c# version suffers less from a need to rewrite in order to use the lambdas without making the lines excessively long even WITH the undesirable indenting.  That makes using lambdas in c# easier.

**Comment by pricerc** [comment]
I'm still not feeling why lambdas are better than LINQ here?

While I can see merit in supporting lambdas, none of the examples I've seen to date would not look better in LINQ (IMO). And you wouldn't need all those inline _Functions_.

E.g. how is:
```VB
For Each a In input.Where(Function(x) x.whatever)
```
better than
```VB
For Each a In (From x In Input Where x.whatever)
```

?

I mean, I'd prefer it this were possible:
```VB
For Each a In Input Where a.whatever
``` 
, but the current state of VB is still more concise and legible than the lambda version.

**Comment by zspitz** [comment]
@pricerc 

> would not look better in LINQ (IMO)

As you say, that is really subjective. I used to feel that the keyword syntax was a better choice, but today I never use it; I still haven't been able to pin down exactly why. Perhaps because it's a level of obfuscation over what the code is actually doing; and I've found myself in enough situations where the abstraction falls down, that I'd rather not have to waste time weighing which one to use.
